tokio/src/io/driver/scheduled_io.rs

```diff
@@ -1,47 +1,30 @@
 use crate::loom::future::AtomicWaker;
 use crate::loom::sync::atomic::AtomicUsize;
-use crate::util::bit;
-use crate::util::slab::{Address, Entry, Generation};
+use crate::util::slab::Entry;
 
-use std::sync::atomic::Ordering::{AcqRel, Acquire, SeqCst};
+use std::sync::atomic::Ordering::{AcqRel, Acquire, Release};
 
+/// Stored in the I/O driver resource slab.
 #[derive(Debug)]
 pub(crate) struct ScheduledIo {
+    /// Packs the resource's readiness with the resource's generation.
     readiness: AtomicUsize,
+
+    /// Task waiting on read readiness
     pub(crate) reader: AtomicWaker,
+
+    /// Task waiting on write readiness
     pub(crate) writer: AtomicWaker,
 }
 
-const PACK: bit::Pack = bit::Pack::most_significant(Generation::WIDTH);
-
 impl Entry for ScheduledIo {
-    fn generation(&self) -> Generation {
-        unpack_generation(self.readiness.load(SeqCst))
-    }
-
-    fn reset(&self, generation: Generation) -> bool {
-        let mut current = self.readiness.load(Acquire);
-
-        loop {
-            if unpack_generation(current) != generation {
-                return false;
-            }
-
-            let next = PACK.pack(generation.next().to_usize(), 0);
+    fn reset(&self) {
+        let state = self.readiness.load(Acquire);
 
-            match self
-                .readiness
-                .compare_exchange(current, next, AcqRel, Acquire)
-            {
-                Ok(_) => break,
-                Err(actual) => current = actual,
-            }
-        }
+        let generation = super::GENERATION.unpack(state);
+        let next = super::GENERATION.pack_lossy(generation + 1, 0);
 
-        drop(self.reader.take_waker());
-        drop(self.writer.take_waker());
-
-        true
+        self.readiness.store(next, Release);
```

Comment:
I guess it"s guaranteed by other code that the state will not be touched between the load and store here, but this is still a little concerning. I have a slight preference for making this a `compare_exchange`, even if it's just so we can `debug_assert!` that it never fails (rather than a cas loop).

Reply:
When it comes to atomics, I prefer to avoid doing work just because it feels safer. IMO, usually, it offers a false sense of security and it is better to focus on understanding the logic (one reason why I *usually* avoid `SeqCst`).

Reply:
That's fair. My thought process was that, because there's nothing _else_ enforcing that the state will not change between the load and the store, it would be good to be able to assert that it hasn't here. But, this is indeed simpler, and that's also worth something. :)

Reply:
It's also worth noting a CAS is *much* heavier than store.